File: python/zigzag-conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s
        rows = []
        for i in range(numRows):
            rows.append([])
        string = [char for char in reversed(s)]
        fill_all = True
        for char in s:
            zigzag_pos = max([len(rows) - 2, 1 if numRows > 2 else 0])
            if fill_all:
                for row in rows:
                    try:
                        row.append(string.pop())
                    except: 
                        break
                fill_all = False if numRows > 2 else True
            else:
                zigzag = True
                try:
                    while zigzag:
                        for index, row in enumerate(rows):
                            if index != zigzag_pos:
                                continue
                            row.append(string.pop())
                            if zigzag_pos == (1 if numRows > 1 else 0):
                                zigzag = False
                            zigzag_pos = max([zigzag_pos - 1, 1 if numRows > 2 else 0])
                    fill_all = True
                except:
                    break
        return "".join(["".join(row) for row in rows])

File: rust/zigzag-conversion.rs
impl Solution {
    pub fn convert(s: String, num_rows: i32) -> String {
        if num_rows == 1 || num_rows > s.len() as i32 {
            return s;
        }
        let mut output = vec![];
        for i in 1..=num_rows {
            output.push("".to_string());
        }
        let mut index: i32 = 0;
        let mut step = 1;
        for letter in s.chars() {
            output[index as usize] += &letter.to_string();
            if index == 0 {
                step = 1;
            } else if index == num_rows - 1 {
                step = -1;
            }
            index += step;
        }
        return output.join("");
    }
}
